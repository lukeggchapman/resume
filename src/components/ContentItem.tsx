import React from 'react'
import { View, Text, StyleSheet, Image } from '@react-pdf/renderer'
import { format } from 'date-fns'

import hr from 'images/hr.png'
import dateDifference from 'utils/dateDifference'
import typography from './typography'

export const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    fontSize: 10,
    padding: '0 5',
  },
  leftColumn: {
    flex: 1,
    alignItems: 'center',
    justifyItems: 'center',
    paddingRight: 5,
  },
  rightColumn: {
    flex: 7,
  },
  subTitle: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  contentContainer: {
    flexDirection: 'column',
    width: '100%',
  },
  content: {
    marginTop: 5,
  },
  logo: {
    height: 35,
  },
  hr: {
    width: 400,
    alignSelf: 'center',
    margin: '10 0',
  },
  dates: {
    flexDirection: 'row',
  },
})

export interface ContentItemProps {
  title: string
  subTitle: string
  logo: string
  startDate: string
  endDate: string
  children?: React.ReactNode
}

const ContentItem = ({
  title,
  subTitle,
  logo,
  startDate,
  endDate,
  children,
}: ContentItemProps) => (
  <View style={styles.container}>
    <View style={styles.leftColumn}>
      <Image style={styles.logo} src={logo} safePath="." />
    </View>
    <View
      style={[
        styles.rightColumn,
        { alignSelf: children ? 'normal' : 'center' },
      ]}
    >
      <Text style={typography.h3}>{title}</Text>
      <View style={[typography.h4, styles.subTitle]}>
        <Text>{subTitle}</Text>
        <View>
          <Text>
            {`${format(startDate, 'MMM, YYYY')} \u2013 ${format(
              endDate,
              'MMM, YYYY'
            )}`}
            {` \u00B7 `}
            {dateDifference(startDate, endDate)}
          </Text>
        </View>
      </View>
      <View style={styles.contentContainer}>
        {children && <View style={styles.content}>{children}</View>}
        <Image style={styles.hr} src={hr} safePath="." />
      </View>
    </View>
  </View>
)

export default ContentItem
