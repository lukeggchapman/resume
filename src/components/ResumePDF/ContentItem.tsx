import React from 'react'
import { View, Text, StyleSheet, Image } from '@react-pdf/renderer'
import { differenceInMonths, addDays, format } from 'date-fns'

import hr from '../../images/hr.png'

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    fontSize: 10,
    padding: '0 5',
  },
  leftColumn: {
    flex: 1,
    alignItems: 'center',
    justifyItems: 'center',
    paddingRight: 5,
  },
  rightColumn: {
    flex: 7,
  },
  title: {
    fontFamily: 'Oswald',
    fontSize: 14,
  },
  subTitle: {
    fontFamily: 'PT Sans',
    fontSize: 10,
    color: '#555',
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  contentContainer: {
    flexDirection: 'column',
    width: '100%',
  },
  content: {
    marginTop: 5,
  },
  logo: {
    height: 35,
  },
  hr: {
    width: 400,
    alignSelf: 'center',
    margin: '10 0',
  },
  dates: {
    flexDirection: 'row',
  },
})

const getDifference = (startDate: string, endDate: string) => {
  const endDateAdd1 = addDays(endDate, 1) // Most contracts will end on the last day of the month
  const total = differenceInMonths(endDateAdd1, startDate)
  const years = Math.floor(total / 12)
  const months = total % 12

  return [
    (years && `${years} yr${years > 1 ? 's' : ''}`) || undefined,
    (months && `${months} mo${months > 1 ? 's' : ''}`) || undefined,
  ]
    .filter(string => !!string)
    .join(' ')
}

interface ContentItemProps {
  title: string
  subTitle: string
  logo?: string
  startDate: string
  endDate: string
  children?: React.ReactNode
}

const ContentItem = ({
  title,
  subTitle,
  logo,
  startDate,
  endDate,
  children,
}: ContentItemProps) => (
  <View style={styles.container}>
    <View style={styles.leftColumn}>
      {logo && <Image style={styles.logo} src={logo} />}
    </View>
    <View
      style={[
        styles.rightColumn,
        { alignSelf: children ? 'normal' : 'center' },
      ]}
    >
      <Text style={styles.title}>{title}</Text>
      <View style={styles.subTitle}>
        <Text>{subTitle}</Text>
        <View>
          <Text>
            {`${format(startDate, 'MMM, YYYY')} \u2013 ${format(
              endDate,
              'MMM, YYYY'
            )}`}
            {` \u00B7 `}
            {getDifference(startDate, endDate)}
          </Text>
        </View>
      </View>
      <View style={styles.contentContainer}>
        {children && <View style={styles.content}>{children}</View>}
        <Image style={styles.hr} src={hr} />
      </View>
    </View>
  </View>
)

export default ContentItem
