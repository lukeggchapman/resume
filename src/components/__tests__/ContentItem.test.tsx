import React from 'react'
import { render } from '@testing-library/react'

import ContentItem, { ContentItemProps } from '../ContentItem'

const baseProps: ContentItemProps = {
  title: 'titleTest',
  subTitle: 'subTitleTest',
  logo: 'logoTest',
  startDate: '2010-05-05',
  endDate: '2012-06-06',
  children: 'ReactNodeTest',
}

describe('ContentItem', () => {
  const renderContentItem = (props?: Partial<ContentItemProps>) =>
    render(<ContentItem {...{ ...baseProps, ...props }} />)

  it('not passing children removes content', () => {
    const contentItem = renderContentItem({ children: undefined })

    expect(contentItem.queryByTestId('content')).toBeNull()
  })

  it('allows setting container style', () => {
    const contentItem = renderContentItem({
      children: undefined,
      style: { padding: '10px' },
    })
    const pdfStyle = JSON.parse(
      (contentItem.container.firstChild as HTMLElement | null)?.dataset.style ??
        '{}'
    )

    expect(pdfStyle).toHaveLength(2)
    expect(pdfStyle[1]).toEqual({ padding: '10px' })
  })
})
