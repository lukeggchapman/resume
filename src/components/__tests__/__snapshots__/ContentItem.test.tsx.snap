// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentItem renders correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "fontSize": 10,
      "padding": "0 5",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "paddingRight": 5,
      }
    }
  >
    <Image
      safePath="."
      src="logoTest"
      style={
        Object {
          "height": 35,
        }
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "flex": 7,
        },
        Object {
          "alignSelf": undefined,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "fontFamily": "Oswald",
          "fontSize": 14,
        }
      }
    >
      titleTest
    </Text>
    <View
      style={
        Array [
          Object {
            "color": "#555",
            "fontFamily": "PT Sans",
            "fontSize": 10,
          },
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <Text>
        subTitleTest
      </Text>
      <View>
        <Text>
          May, 2010 – Jun, 2012
           · 
          2 yrs 1 mo
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "marginTop": 5,
          }
        }
      >
        ReactNodeTest
      </View>
      <Image
        safePath="."
        src="test-file-stub"
        style={
          Object {
            "alignSelf": "center",
            "margin": "10 0",
            "width": 400,
          }
        }
      />
    </View>
  </View>
</View>
`;
